src/kaleidoscope/driver/scanner/ATMegaScannerDescription.h

```diff
@@ -0,0 +1,68 @@
+ * Copyright (C) 2018-2019  Keyboard.io, Inc
+ *
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, version 3.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <Arduino.h>
+
+#include "kaleidoscope/driver/ScannerDescription.h"
+#include "kaleidoscope/hardware/avr/pins_and_ports.h"
+#include "kaleidoscope/macro_helpers.h"
+
+#define ROW_PIN_LIST(...)  __VA_ARGS__
+#define COL_PIN_LIST(...)  __VA_ARGS__
+
+#define ATMEGA_SCANNER_DESCRIPTION(ROW_PINS_, COL_PINS_)                \
+  SCANNER_DESCRIPTION(NUM_ARGS(ROW_PINS_), NUM_ARGS(COL_PINS_));        \
+  static constexpr uint8_t matrix_row_pins[matrix_rows] =  ROW_PINS_;   \
+  static constexpr uint8_t matrix_col_pins[matrix_columns] =  COL_PINS_;
+
+#define ATMEGA_SCANNER_DESCRIPTION_DATA(BOARD)              \
+  SCANNER_DESCRIPTION_DATA(BOARD);                          \
+  constexpr uint8_t BOARD::matrix_row_pins[matrix_rows];    \
+  constexpr uint8_t BOARD::matrix_col_pins[matrix_columns];
+
+#define _SD(BOARD) \
+  BOARD::ScannerDescription_
+#define ATMEGA_SCANNER_DATA(BOARD)                                                         \
+  ATMEGA_SCANNER_DESCRIPTION_DATA(BOARD ## Description)                                    \
+  template<>                                                                               \
+  volatile uint16_t BOARD::previousKeyState_[_SD(BOARD).matrix_rows] = {};                 \
+  template<>                                                                               \
+  volatile uint16_t BOARD::keyState_[_SD(BOARD).matrix_rows] = {};                         \
+  template<>                                                                               \
+  uint16_t BOARD::masks_[_SD(BOARD).matrix_rows] = {};                                     \
+  template<>                                                                               \
+  uint8_t BOARD::debounce_matrix_[_SD(BOARD).matrix_rows][_SD(BOARD).matrix_columns] = {}; \
+                                                                                           \
+  ISR(TIMER1_OVF_vect) {                                                                   \
+    KeyboardHardware.readMatrix();                                                         \
+  }
+
+namespace kaleidoscope {
+namespace driver {
+namespace scanner {
+
+struct ATMegaScannerDescription : public kaleidoscope::driver::ScannerDescription {
+  static constexpr uint8_t matrix_row_pins[] = {};
```

Comment:
I suppose I can add some comments explaining what this is for.

Reply:
You and C++ surprised me that this is legal code. Only checking with compiler explorer and latest gcc I figured that it must be legal, although I have no clue what this is supposed to mean from a language point of view. An empty array?

Reply:
Yep, an empty array. Missing elements are filled with zeroes (since it's static), which is just fine for us here. The single purpose of these arrays is to please the compiler, every single derivative will override them.

Reply:
> Missing elements are filled with zeroes (since it's static), which is just fine for us here.

Nope. There are no zeroes as the respective array is empty. And that's what confuses me. Mainly because it is legal to take the address of the array even for constexpr arrays (mind you, those can but do not have to be compile time constant).

The following code compiles with gcc and -Wall -Werror without any warnings or errors.
```cpp
struct Test {
    static constexpr int a[] = {};
};

int main() {
    const int *test = Test::a;
    return test[0];
}
```
Only when you add the flag `-pedantic`, the [compiler](https://www.godbolt.org/z/CkbxLE) decends to emitting an error 
```
error: zero-size array 'Test::a'

static constexpr int a[] = {};

                     ^
```

Adding `static_assert(sizeof(Test::a) == 0, "");` shows that the compiler regards the array as actually empty. The question that remains to be answered is to which address does the pointer `Test::a` point to?

Reply:
Right... I was confusing this with the derived classes that specify a size for the array. Then it would initialize to zeroes. But you're right, in this case this is just an empty array.

As for where it points to, I don't think it matters. Noone should be using `ATMegaScanner::matrix_row_pins` directly, derived classes should override those. As such, I'm not fussed about what happens with the base struct's members, noone should be using them.